flintrock/core.py

```diff
@@ -482,6 +483,58 @@ def _run_asynchronously(*, partial_func: functools.partial, hosts: list):
         loop.close()
 
 
+def get_java_major_version(client: paramiko.client.SSHClient):
+    possible_cmds = [
+        "$JAVA_HOME/bin/java -version 2>&1",
+        "java -version 2>&1"
+    ]
+
+    for command in possible_cmds:
+        try:
+            output = ssh_check_output(
+                client=client,
+                command=command)
+            tokens = output.split()
+            # First line of the output is like: 'java version "1.8.0_20"'
+            # Get the version string and strip out the first two parts of the
+            # version as a tuple: (1, 8)
+            if len(tokens) >= 3:
+                version_parts = tokens[2].strip('"').split(".")
+                if len(version_parts) >= 2:
+                    return tuple(int(part) for part in version_parts[:2])
+        except SSHError:
+            pass
+
+    return None
+
+
+def ensure_java8(client: paramiko.client.SSHClient):
+    host = client.get_transport().getpeername()[0]
+    java_major_version = get_java_major_version(client)
+
+    # The default CentOS AMIs on EC2 don't come with Java installed.
+    # Amazon Linux only have Java 1.7 installed.
+    if not java_major_version or java_major_version < (1, 8):
+        print("[{h}] Installing Java 1.8...".format(h=host))
+
+        ssh_check_output(
+            client=client,
+            command="""
+                set -e
+
+                # Install Java 1.8 first to protect packages that depend on Java from being removed.
+                sudo yum install -y java-1.8.0-openjdk
+
+                # Remove any older versions of Java to force the default Java to 1.8.
+                # We don't use alternatives because it does not seem to update links in /usr/lib/jvm correctly,
```

Comment:
Can we update this to say "We don't use /etc/alternatives..." or something like that so it's clear that we're talking about a specific piece software? Otherwise I will probably forget what this is about in a few months. 😆